src/TodoContext.tsx

```diff
@@ -0,0 +1,304 @@
+
+          newTodos.splice(index, 1, todo);
+
+          return newTodos;
+        });
+
+        return true;
+      } catch {
+        handleErrorMessageSend(Errors.update);
+
+        return false;
+      } finally {
+        setIsLoading(prev => prev.filter(todoId => todoId !== id));
+      }
+    },
+    [handleErrorMessageSend],
+  );
+
+  const handleToggleComplete = useCallback(
+    async (updatedTodo: Todo) => {
+      setIsLoading(prev => [...prev, updatedTodo.id]);
+
+      try {
+        const currentTodo = await todoService.updateTodo({
+          ...updatedTodo,
+          completed: !updatedTodo.completed,
+        });
+
+        setTodos(prevTodos =>
+          prevTodos.map(todo =>
+            todo.id === currentTodo.id ? currentTodo : todo,
+          ),
+        );
+      } catch {
+        handleErrorMessageSend(Errors.update);
+      } finally {
+        setIsLoading(prev => prev.filter(todoId => todoId !== updatedTodo.id));
+      }
+    },
+    [handleErrorMessageSend],
+  );
+
+  const handleToggleAllComplete = useCallback(
+    async (updatedTodos: Todo[]) => {
+      const isAllCompleted = updatedTodos.every(t => t.completed);
+
+      const tempArrTodos = isAllCompleted
+        ? updatedTodos
+        : updatedTodos.filter(t => !t.completed);
+
+      for (const todo of tempArrTodos) {
+        setIsLoading(prev => [...prev, todo.id]);
+
+        try {
+          const currentTodo = await todoService.updateTodo({
+            ...todo,
+            completed: !todo.completed,
+          });
+
+          setTodos(prevTodos =>
+            prevTodos.map(t => (t.id === currentTodo.id ? currentTodo : t)),
+          );
+        } catch {
+          handleErrorMessageSend(Errors.update);
+        } finally {
+          setIsLoading(prev => prev.filter(todoId => todoId !== todo.id));
+        }
+      }
+    },
+    [handleErrorMessageSend],
+  );
+
+  const todosValue = useMemo(
+    () => ({
+      todos,
+      filteredTodos,
+      tempTodo,
+      filterStatus,
+      isSubmitting,
+    }),
+    [todos, filteredTodos, tempTodo, filterStatus, isSubmitting],
+  );
+
+  const apiValue = useMemo(
+    () => ({
+      handleCreateTodo,
+      handleDeleteTodo,
+      handleUpdateTodo,
+      handleToggleComplete,
+      handleToggleAllComplete,
+      handleChangeFilterStatus,
+    }),
+    [
+      handleCreateTodo,
+      handleDeleteTodo,
+      handleUpdateTodo,
+      handleToggleComplete,
+      handleToggleAllComplete,
+      handleChangeFilterStatus,
+    ],
+  );
+
+  useEffect(() => {
+    const fetchTodos = async () => {
+      try {
+        const loadedTodos = await getTodos();
+
+        setTodos(loadedTodos);
+      } catch {
+        handleErrorMessageSend(Errors.load);
+      }
+    };
+
+    fetchTodos();
+  }, []);
+
+  return (
+    <TodosApiContext.Provider value={apiValue}>
+      <TodosLoadingContext.Provider value={isLoading}>
+        <TodosContext.Provider value={todosValue}>
```

Comment:
I think that it isn't a good idea to create 3 different providers and it will be better to combine all your logic in one provider